File: scm-csharp/scheme/primitives/PrimitiveDirectoryName.cs
namespace scheme;

public class PrimitiveDirectoryName : Primitive
{
    public override string Name()
    {
        return "directory-name";
    }

    public override string Info()
    {
        return
            "Syntax: (directory-name path)\n" +
            "Library: (scm system)\n" +
            "Description: Returns the directory part of the given path as an absolute path string, or #f if there is no parent directory.\n" +
            "Example:\n" +
            "  (directory-name \"/usr/share/readme.txt\") => \"/usr/share\"";
    }
    
    public override object Apply(SourcePos? pos, object[] arguments)
    {
        CheckArgs(pos, arguments, 1, 1);
        var path = new String(Value.AsString(arguments[0]));
        var dirinfo = Directory.GetParent(path);
        if (dirinfo == null) return Value.F;
        return dirinfo.FullName.ToCharArray();
    }
}

File: scm-java/src/main/scheme/primitives/PrimitiveDirectoryName.java
package scheme.primitives;

import java.io.File;

import scheme.*;

public class PrimitiveDirectoryName extends Primitive {
    @Override
    public String name() {
        return "directory-name";
    }

    @Override
    public String info() {
        return "Syntax: (directory-name path)\n" +
               "Library: (scm fs)\n" +
               "Description: Returns the directory part of the given path as an absolute path string, or #f if there is no parent directory.\n" +
               "Example:\n" +
               "  (directory-name \"/usr/share/readme.txt\") => \"/usr/share\"";
    }
    
    @Override
    public Object apply(SourcePos pos, Object[] arguments) {
        checkArgs(pos, arguments, 1, 1);
        var path = new File(new String(Value.asString(arguments[0])));
        path = new File(path.getAbsolutePath());
        return path.getParent().toCharArray();
    }
}
